controllers/clustercache/cluster_cache.go

```diff
@@ -0,0 +1,650 @@
+// A source.Kind source (configured with Kind, EventHandler and Predicates) will be added to the Watcher.
+// To watch for events, the source.Kind will create an informer on the Cache that we have created and cached
+// for the given Cluster.
+type WatchInput struct {
+	// Name represents a unique Watch request for the specified Cluster.
+	// The name is used to track that a specific watch is only added once to a cache.
+	// After a connection (and thus also the cache) has been re-created, watches have to be added
+	// again by calling the Watch method again.
+	Name string
+
+	// Cluster is the object key of the workload cluster.
+	Cluster client.ObjectKey
+
+	// Watcher is the watcher (controller) whose Reconcile() function will be called for events.
+	Watcher Watcher
+
+	// Kind is the type of resource to watch.
+	Kind client.Object
+
+	// EventHandler contains the event handlers to invoke for resource events.
+	EventHandler handler.EventHandler
+
+	// Predicates is used to filter resource events.
+	Predicates []predicate.Predicate
+}
+
+// GetClusterSourceOption is an option that modifies GetClusterSourceOptions for a GetClusterSource call.
+type GetClusterSourceOption interface {
+	// ApplyToGetClusterSourceOptions applies this option to the given GetClusterSourceOptions.
+	ApplyToGetClusterSourceOptions(option *GetClusterSourceOptions)
+}
+
+// GetClusterSourceOptions allows to set options for the GetClusterSource method.
+type GetClusterSourceOptions struct {
+	watchForProbeFailures []time.Duration
+}
+
+// ApplyOptions applies the passed list of options to GetClusterSourceOptions,
+// and then returns itself for convenient chaining.
+func (o *GetClusterSourceOptions) ApplyOptions(opts []GetClusterSourceOption) *GetClusterSourceOptions {
+	for _, opt := range opts {
+		opt.ApplyToGetClusterSourceOptions(o)
+	}
+	return o
+}
+
+// WatchForProbeFailure will configure the Cluster source to send an event if the health probe
+// didn't succeed for the configured duration.
+// For example if WatchForProbeFailure is set to 5m, an event will be sent if LastProbeSuccessTimestamp
+// is 5m in the past (i.e. health probes didn't succeed in the last 5m).
+type WatchForProbeFailure time.Duration
+
+// ApplyToGetClusterSourceOptions applies WatchForProbeFailure to the given GetClusterSourceOptions.
+func (n WatchForProbeFailure) ApplyToGetClusterSourceOptions(opts *GetClusterSourceOptions) {
+	opts.watchForProbeFailures = append(opts.watchForProbeFailures, time.Duration(n))
+}
+
+// SetupWithManager sets up a ClusterCache with the given Manager and Options.
+// This will add a reconciler to the Manager and returns a ClusterCache which can be used
+// to retrieve e.g. Clients for a given Cluster.
+func SetupWithManager(ctx context.Context, mgr manager.Manager, options Options, controllerOptions controller.Options) (ClusterCache, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("controller", "clustercache")
```

Comment:
I think it's a good time to change the name:
1. I think the Tracker suffix doesn't make sense (or at least I have no idea what it stands for)
2. It's a good way to signal that this is a new iteration of the component 
3. 99% of CAPI contributors or users don't know that the ClusterCacheTracker exists

I think it's easier to talk about old ClusterCacheTracker vs new ClusterCache instead of old/new ClusterCacheTracker.